refactor(service-worker): improve typings/mocks and switch to official TypeScript typings

gkalpak opened the pull request:
This PR includes several improvements to the ServiceWorker typings and mocks used in tests, bringing the typings and mock implementations closer to the actual implementations. It also switches from using custom typings for ServiceWorker APIs to using the official TypeScript ones.

See individual commits for details.

##
The PR also includes a small fix (related to notifying clients about an unrecoverable state), which was revealed by the more accurate typings.

See the last commit for more details.

mary-poppins commented:
You can preview 9a28877 at https://pr42736-9a28877.ngbuilds.io/.

mary-poppins commented:
You can preview 316dc2f at https://pr42736-316dc2f.ngbuilds.io/.

gkalpak commented on packages/localize/src/localize/src/global.ts:
```diff
@@ -10,7 +10,13 @@
 
 declare global {
-  var WorkerGlobalScope: any;
+  interface WorkerGlobalScope extends EventTarget, WindowOrWorkerGlobalScope {
+  }
+
+  var WorkerGlobalScope: {
+    prototype: WorkerGlobalScope;
+    new(): WorkerGlobalScope;
+  };
```
This change is needed, because all files under `packages/` are compiled together as part of the [legacy-unit-tests-saucelabs][1] CI job, including the `lib.webworker.d.ts` typings brought in by [service-worker/worker/src/service-worker.d.ts][2].

[1]: https://github.com/angular/angular/blob/ffeea63f43e6a7fd46be4a8cd5a5d254c98dea08/.circleci/config.yml#L681
[2]: https://github.com/angular/angular/blob/316dc2f12ce8931f5ff66fa5f8da21c0d251a337/packages/service-worker/worker/src/service-worker.d.ts#L9

gkalpak commented on packages/router/test/bootstrap.spec.ts:
```diff
@@ -13,6 +13,8 @@ import {BrowserModule} from '@angular/platform-browser';
 import {platformBrowserDynamic} from '@angular/platform-browser-dynamic';
 import {NavigationEnd, Resolve, Router, RouterModule} from '@angular/router';
 
+declare var window: Window;
```
This change is needed, because all files under `packages/` are compiled together as part of the [legacy-unit-tests-saucelabs][1] CI job, including the `lib.webworker.d.ts` typings brought in by [service-worker/worker/src/service-worker.d.ts][2].

[1]: https://github.com/angular/angular/blob/ffeea63f43e6a7fd46be4a8cd5a5d254c98dea08/.circleci/config.yml#L681
[2]: https://github.com/angular/angular/blob/316dc2f12ce8931f5ff66fa5f8da21c0d251a337/packages/service-worker/worker/src/service-worker.d.ts#L9

gkalpak replied:
We can :smiley:

gkalpak commented on packages/localize/src/localize/src/global.ts:
```diff
@@ -10,7 +10,9 @@
 
 declare global {
-  var WorkerGlobalScope: any;
+  interface WorkerGlobalScope extends EventTarget, WindowOrWorkerGlobalScope {}
+
+  var WorkerGlobalScope: {prototype: WorkerGlobalScope; new (): WorkerGlobalScope;};
```
This change is needed, because all files under `packages/` are compiled together as part of the [legacy-unit-tests-saucelabs][1] CI job, including the `lib.webworker.d.ts` typings brought in by [service-worker/worker/src/service-worker.d.ts][2].

[1]: https://github.com/angular/angular/blob/ffeea63f43e6a7fd46be4a8cd5a5d254c98dea08/.circleci/config.yml#L681
[2]: https://github.com/angular/angular/blob/316dc2f12ce8931f5ff66fa5f8da21c0d251a337/packages/service-worker/worker/src/service-worker.d.ts#L9

gkalpak replied:
Added.

mary-poppins commented:
You can preview bd32b3b at https://pr42736-bd32b3b.ngbuilds.io/.

atscott reviewed (approved):
reviewed-for: fw-router

AndrewKushnir commented on packages/service-worker/worker/testing/events.ts:
```diff
@@ -0,0 +1,124 @@
+}
+
+export class MockExtendableEvent extends MockEvent implements ExtendableEvent {
+  private queue: Promise<unknown>[] = [];
```
Would it make sense to add a `super` call here to match the logic of other classes in this file?

gkalpak replied:
What arguments would be passed to it? Unlike the other classes (which pass a specific `type` argument to their parent constructors), this is mostly used as a base class and there is not specific type to pass to `super()`.

Am I missing something?

AndrewKushnir reviewed (approved):
LGTM (just left a few minor comments), great that we are switching to the official typings! 👍

gkalpak committed:
refactor(service-worker): remove unused variables and imports from tests

This commit removes some unused variables (and associated imports) from
`integration_spec.ts`.

gkalpak committed:
refactor(service-worker): make `SwTestHarness.envIsSupported()` a standalone function

This commit makes the `SwTestHarness.envIsSupported()` static method a
standalone function. This function is used to determine whether the
current environment provides the necessary APIs to run the SW tests and
is independent of `SwTestHarness`, so is no need for it to be a static
method of `SwTestHarness`.

This is in preparation of switching from our custom typings to the
official TypeScript typings (`lib.webworker.d.ts`).

gkalpak committed:
refactor(service-worker): move mock event classes to their own file

In the ServiceWorker tests, we use mock implementations of the various
events emitted during the ServiceWorker lifecycle. Previously, these
mock implementations were defined in the `testing/scope.ts` file. This
added several extra classes to a file that already contains a few,
making it harder to maintain.

Therefore, this commit moves these mock event classes to a separate
`testing/events.ts` file.

gkalpak committed:
test(service-worker): better align mock event implementations with actual implementations

This commit better aligns the mock event implementations used in
ServiceWorker tests (and the associated typings) with the actual
implementations (and the official TypeScript typings). This allows
verifying the ServiceWorker behavior in a slightly more realistic
environment.

This is in preparation of switching from our custom typings to the
official TypeScript typings (`lib.webworker.d.ts`).

gkalpak committed:
refactor(service-worker): remove duplicate `Context` type (in favor of `ExtendableEvent`)

This commit removes the duplicate `Context` interface and uses the
`ExtendableEvent` interface instead.

This is in preparation of switching from our custom typings to the
official TypeScript typings (`lib.webworker.d.ts`).

gkalpak committed:
refactor(service-worker): move mock client classes to their own file

In the ServiceWorker tests, we use mock implementations of the various
client APIs that the ServiceWorker interacts with. Previously, these
mock implementations were defined in the `testing/scope.ts` file. This
added several extra classes to a file that already contains a few,
making it harder to maintain.

Therefore, this commit moves these mock client classes to a separate
`testing/clients.ts` file.

gkalpak committed:
test(service-worker): better align mock global scope implementation with actual implementation

This commit better aligns the mock `ServiceWorkerGlobalScope`
implementation used in ServiceWorker tests (and the associated typings)
with the actual implementation (and the official TypeScript typings).
This allows verifying the ServiceWorker behavior in a slightly more
realistic environment.

This is in preparation of switching from our custom typings to the
official TypeScript typings (`lib.webworker.d.ts`).

gkalpak committed:
test(service-worker): better align mock client implementations with actual implementations

This commit better aligns the mock client implementations used in
ServiceWorker tests (and the associated typings) with the actual
implementations (and the official TypeScript typings). This allows
verifying the ServiceWorker behavior in a slightly more realistic
environment.

This is in preparation of switching from our custom typings to the
official TypeScript typings (`lib.webworker.d.ts`).

gkalpak committed:
refactor(service-worker): switch to the official TypeScript typings

Previously, we used custom typings for the ServiceWorker environment.
This was necessary back when the ServiceWorker package was introduced,
since there were no official typings.

Since there are now official typings for Web Workers (including
ServiceWorkers) offered by TypeScript as [lib.webworker.d.ts][1], this
commit gets rid of our custom typings in favor of using the official
ones.

[1]: https://github.com/microsoft/TypeScript/blob/v4.3.4/lib/lib.webworker.d.ts

gkalpak committed:
fix(service-worker): correctly handle unrecoverable state when a client no longer exists

Previously, the ServiceWorker assumed that a client found in
`clientVersionMap` would exist (i.e. it could be retrieved via
`clients.get()`). However, if a browser tab had been closed, the
corresponding client (while present in `clientVersionMap`, which is only
updated on ServiceWorker initialization) would not be retrievable via
`clients.get()`.

This commit fixes it by checking whether the client exists before trying
to notify it about an unrecoverable state.

gkalpak committed:
fixup! test(service-worker): better align mock client implementations with actual implementations

gkalpak commented on packages/service-worker/worker/testing/clients.ts:
```diff
@@ -0,0 +1,96 @@
+        (type === 'all') ? allClients : allClients.filter(client => client.type === type);
+
+    // Order clients in most recently focused order to adhere to the spec.
+    return matchedClients.reverse().sort((a, b) => b.lastFocusedAt - a.lastFocusedAt) as any;
```
Yes, it's about return clients in most recently created order when `lastFocusedAt` is the same. I'll add a comment.

gkalpak commented on packages/service-worker/worker/testing/clients.ts:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  async get(id: string): Promise<Client> {
+    return this.clients.get(id)! as any as Client;
```
We don't need both indeed. I kept the non-null assertion and removed the type casting.

gkalpak committed:
fixup! refactor(service-worker): switch to the official TypeScript typings

mary-poppins commented:
You can preview fe02813 at https://pr42736-fe02813.ngbuilds.io/.

gkalpak commented:
Thx for the reviews, @atscott and @AndrewKushnir 👍
I've addressed the review comments. Marking for merging.
